### PR TITLE
[Profiler] Fix parsing thread information

### PR DESCRIPTION
**Pull request by chrisnas**
## Summary of changes
Fix parsing of /proc/self/task/%d/stat to get thread running state and cpu consumption

## Reason for change
Test were failing in .NET 6

## Implementation details
Take into account thread names with SPACE and ()

### CODE REVIEW

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/OsSpecificApi.cpp [hunk 1]
```diff
@@ -60,10 +69,10 @@ bool GetCpuInfo(pid_t tid, bool& isRunning, uint64_t& cpuTime)
     char state = ' ';   // 3rd position  and 'R' for Running
     int userTime = 0;   // 14th position in clock ticks
     int kernelTime = 0; // 15th position in clock ticks
-    bool success =
-        fscanf(file, "%*s %*s %c %*s %*s %*s %*s %*s %*s %*s %*s %*s %*s %d %d",
-               &state, &userTime, &kernelTime) == 3;
-    fclose(file);
+
+    auto pos = sline.find_last_of(")");
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/OsSpecificApi.cpp:73]: for the question (I asked it myself): are parenthesis always used ?
answerd: This is the code in the kernel that write in that file and there will always be parenthesis https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/proc/array.c#n567

**Review by gleocadie — APPROVED** [review]

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/OsSpecificApi.cpp [hunk 2]
```diff
@@ -50,8 +53,14 @@ bool GetCpuInfo(pid_t tid, bool& isRunning, uint64_t& cpuTime)
 {
     char statPath[64];
     snprintf(statPath, sizeof(statPath), "/proc/self/task/%d/stat", tid);
-    FILE* file = fopen(statPath, "r");
-    if (file == nullptr)
+
+    // load the line to be able to parse it in memory
+    std::ifstream file;
+    file.open(statPath);
```
Comment (andrewlock) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/OsSpecificApi.cpp:59]: Previously it looks like `fopen` may possible fail `(file == nullptr)` is that no longer the case?

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/OsSpecificApi.cpp [hunk 3]
```diff
@@ -60,10 +69,10 @@ bool GetCpuInfo(pid_t tid, bool& isRunning, uint64_t& cpuTime)
     char state = ' ';   // 3rd position  and 'R' for Running
     int userTime = 0;   // 14th position in clock ticks
     int kernelTime = 0; // 15th position in clock ticks
-    bool success =
-        fscanf(file, "%*s %*s %c %*s %*s %*s %*s %*s %*s %*s %*s %*s %*s %d %d",
-               &state, &userTime, &kernelTime) == 3;
-    fclose(file);
+
+    auto pos = sline.find_last_of(")");
+    const char* pEnd = sline.c_str() + pos + 1;
```
Comment (kevingosse) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/OsSpecificApi.cpp:74]: I believe you should check that `pos + 1` is within the string

**Review by kevingosse — COMMENTED** [review]
I think it would be nice to have a unit test for this

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/OsSpecificApi.cpp [hunk 4]
```diff
@@ -60,10 +69,10 @@ bool GetCpuInfo(pid_t tid, bool& isRunning, uint64_t& cpuTime)
...
+    bool success = sscanf(pEnd, " %c %*s %*s %*s %*s %*s %*s %*s %*s %*s %*s %d %d", &state, &userTime, &kernelTime) == 3;
```
[the remark this answers is not in the document]
Comment (andrewlock) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/OsSpecificApi.cpp:75]: image

**Comment by andrewlock** [comment]
[Woohoo, it fixes the integration tests](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=103531&view=results)! 🎉

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/OsSpecificApi.cpp (hunk 2 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/OsSpecificApi.cpp:61]: > Previously it looks like `fopen` may possible fail `(file == nullptr)` is that no longer the case?

It is no more the case: the line will be empty (no exception)

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/OsSpecificApi.cpp (hunk 3 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/OsSpecificApi.cpp:76]: if final ), then points to the \0

**Comment by chrisnas** [comment]
> [Woohoo, it fixes the integration tests](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=103531&view=results)! 🎉

We are doing that just for you @andrewlock ;^)

**Review by kevingosse — APPROVED** [review]

**Pull request merged**